packages/cube-frontend-web-app/src/pages/node/[name]/_components/useIPMIOperations.ts

```
@@ -0,0 +1,73 @@
+import {
+  Node,
+  NodeStatusEnum,
+  OperateNodeIpmiOperationEnum,
+} from '@cube-frontend/api'
+import { nodesApi } from '@cube-frontend/web-app/api/cosApi'
+import { DataCenterContext } from '@cube-frontend/web-app/context/DataCenterContext'
+import { useContext, useState } from 'react'
+
+type UseIPMIOperations = {
+  isIPMIOperating: boolean
+  isInPoweringStatus: boolean
+  showPowerOn: boolean
+  showPowerOff: boolean
+  showPowerCycle: boolean
+  onPowerOnClick: () => Promise<void>
+  onPowerOffClick: () => Promise<void>
+  onPowerCycleClick: () => Promise<void>
+}
+
+const poweringStatuses = new Set<NodeStatusEnum>([
+  NodeStatusEnum.PoweringOn,
+  NodeStatusEnum.PoweringOff,
+  NodeStatusEnum.PoweringCycle,
+])
+
+export const useIPMIOperations = (
+  node: Node | undefined,
+): UseIPMIOperations => {
+  const { dataCenter } = useContext(DataCenterContext)
+
+  const [isIPMIOperating, setIsIPMIOperating] = useState(false)
+
+  const operate = async (
+    operation: OperateNodeIpmiOperationEnum,
+  ): Promise<void> => {
+    setIsIPMIOperating(true)
+    try {
+      await nodesApi.operateNodeIpmi({
+        dataCenter: dataCenter!.name,
+        nodeName: node?.hostname ?? '',
```

Discussion:
Maybe we can throw an error or early return if `nodeName` is `null` or an empty string before the API call.

---

Absolutely. Now that you mention it, the non-null assertion operator `!` also looks reasonable to me, just like how we did it with `dataCenter!.name`. Do you think it make sense to use `node!.hostname` here? (it is to me)

---

Sure!